File: 01-Programming-Languages/C++/Algorithms/Searching/binary_search.cpp
/*
Problem: Binary Search
Given a sorted array of integers, find the index of a target value using binary search.

Approach:
- Divide and conquer, compare target with middle element

Time Complexity: O(log n)
Space Complexity: O(1)

Example:
Input: arr = [1, 3, 5, 7, 9, 11, 13], target = 7
Output: 3
*/

#include <iostream>
#include <vector>
using namespace std;

int binary_search(const vector<int>& arr, int target)
{
    int left = 0, right = arr.size() - 1;
    while (left <= right)
    {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target) return mid;
        if (arr[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return -1;
}

int main()
{
    vector<int> arr = {1, 3, 5, 7, 9, 11, 13};
    int target = 7;
    int result = binary_search(arr, target);
    cout << "Index of " << target << ": " << result << endl;
    return 0;
}

File: 01-Programming-Languages/Scala/Algorithms/Searching/BinarySearch.scala
// Problem: Binary Search
// Description: Find the index of a target value in a sorted array.
//
// Approach:
// - Repeatedly divide the search interval in half
// - Compare target with middle element to narrow the search
//
// Time Complexity: O(log n)
// Space Complexity: O(1)
//
// Example:
// Input: arr = Array(1, 3, 5, 7, 9, 11), target = 7
// Output: 3

object BinarySearch {
  def search(arr: Array[Int], target: Int): Int = {
    var left = 0; var right = arr.length - 1
    while (left <= right) {
      val mid = left + (right - left) / 2
      if (arr(mid) == target) return mid
      if (arr(mid) < target) left = mid + 1
      else right = mid - 1
    }
    -1
  }

  def main(args: Array[String]): Unit = {
    val arr = Array(1, 3, 5, 7, 9, 11)
    val target = 7
    val idx = search(arr, target)
    println(s"Array: ${arr.mkString(", ")}")
    println(s"Target: $target -> Index: $idx")
  }
}
